### PR TITLE
Port WAMR to ESP-IDF

### LINKED ISSUE #883: Build for esp32 fails
Hi there! First of all thanks for putting together WAMR! 
I am trying to build the demo in `product-mini/platforms/esp-idf`, but get an error concerning the headers in `core/shared/platform/esp-idf`.
When I execute the following in my build folder:
```
cmake .. -DCMAKE_TOOLCHAIN_FILE=$IDF_PATH/tools/cmake/toolchain-esp32.cmake -DTARGET=esp32 -GNinja
cmake --build .
```

I get the error
```
/../wasm-micro-runtime/product-mini/platforms/esp-idf/../../../core/shared/platform/esp-idf/platform_internal.h:18:10: fatal error: FreeRTOS.h: No such file or directory
 #include <FreeRTOS.h>
          ^~~~~~~~~~~~
compilation terminated.
```

When trying to resolve this by putting `freertos/FreeRTOS.h`, I end up with the following error:
```
/platform/esp-idf/platform_internal.h:21:10: fatal error: os_api.h: No such file or directory
 #include <os_api.h>
          ^~~~~~~~~~
compilation terminated.
```

I neither found the header `os_api.h` in WAMR nor in ESP-IDF.

### PR DESCRIPTION
**Pull request by 1c3t3a**
This PR introduces an implementation of the WAMR platform API for ESP-IDF and enables support for Espressif microcontrollers.

This PR is related to the following issues at WAMR: closes #883, #628, #449 and #668 as well as [#4735](https://github.com/espressif/esp-idf/issues/4735) at the esp-idf repo.
This PR implements most functions required by [platform_api_vmcore.h](https://github.com/bytecodealliance/wasm-micro-runtime/blob/bbaf0a3c37378d9b6e14bb6e216b956bf1eca10b/core/shared/platform/include/platform_api_vmcore.h) only the following functions are left empty:
- os_thread_get_stack_boundary
- os_mprotect
- os_dcache_flush (although optional)

Furthermore, a lot of functions from [platform_api_extension.h](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/shared/platform/include/platform_api_extension.h) are implemented. Only the following functions are missing:
- os_recursive_mutex_init
- os_thread_env_init
- os_thread_env_destroy

It would be nice to get feedback from someone more familiar with ESP-IDF on how to implement those functions. I am working with ESP-IDF mostly through it's Rust API and I currently don't feel safe implementing these methods without a mentor or someone that looks over my work.

The code in this PR runs on my Esp32c3 and Esp32 in Interpreter mode, the AoT mode currently errors on both platforms with a `Guru Meditation Error` I think that is because `os_mprotect` isn't implemented. 
I also added some documentation around how to build WAMR for ESP-IDF.

### CODE REVIEW

**Comment by sfranzyshen** [comment]
👍

File: core/shared/platform/esp-idf/espidf_thread.c [hunk 1]
```diff
@@ -3,12 +3,167 @@
...
+int
+os_cond_reltimedwait(korp_cond *cond, korp_mutex *mutex, uint64 useconds)
+{
+    int ret;
+    struct timespec abstime;
+
+    if (useconds == BHT_WAIT_FOREVER)
+        ret = pthread_cond_wait(cond, mutex);
+    else {
+        msec_nsec_to_abstime(&abstime, useconds);
```
Comment (wenyongh) [inline core/shared/platform/esp-idf/espidf_thread.c:155]: msec_nsec_to_abstime isn't implemented?

File: doc/build_wamr.md [hunk 2]
```diff
@@ -457,6 +460,18 @@ NuttX
...
+-------------------------
+WAMR integrates with ESP-IDF both for the XTENSA and RISC-V chips (esp32x and esp32c3 respectively). 
+
+In order to use this, you need version at least version 4.3.1 of ESP-IDF. 
+If you don't have it installed, follow the instructions [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/#get-started-get-prerequisites).
+ESP-IDF also installs the toolchains needed for compiling WAMR and ESP-IDF.
+A small demonstration of how to use WAMR and ESP-IDF can be found under [product_mini](/product-mini/platforms/esp-idf).
+The demo builds WAMR for ESP-IDF and runs a small wasm program. 
+In order to run it for you're specific Espressif chip, edit the ['build.sh'](/product-mini/platforms/esp-idf/build.sh) file and put the correct toolchain file (see #Cross-compilation) and `IDF_TARGET`.
+Before compiling it is also necessary to calls ESP-IDF's `export.sh` script to bring all compile time relevant information in scope.
```
Comment (wenyongh) [inline doc/build_wamr.md:473]: calls -> call

File: doc/build_wamr.md [hunk 3]
```diff
@@ -457,6 +460,18 @@ NuttX
 -------------------------
 WAMR is intergrated with NuttX, just enable the WAMR in Kconfig option (Application Configuration/Interpreters).
 
+ESP-IDF
+-------------------------
+WAMR integrates with ESP-IDF both for the XTENSA and RISC-V chips (esp32x and esp32c3 respectively). 
+
+In order to use this, you need version at least version 4.3.1 of ESP-IDF.
```
Comment (wenyongh) [inline doc/build_wamr.md:467]: How about `you need at least version 4.3.1 of ESP-ID`

File: doc/build_wamr.md [hunk 4]
```diff
@@ -457,6 +460,18 @@ NuttX
...
+ESP-IDF
+-------------------------
+WAMR integrates with ESP-IDF both for the XTENSA and RISC-V chips (esp32x and esp32c3 respectively). 
+
+In order to use this, you need version at least version 4.3.1 of ESP-IDF. 
+If you don't have it installed, follow the instructions [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/get-started/#get-started-get-prerequisites).
+ESP-IDF also installs the toolchains needed for compiling WAMR and ESP-IDF.
+A small demonstration of how to use WAMR and ESP-IDF can be found under [product_mini](/product-mini/platforms/esp-idf).
+The demo builds WAMR for ESP-IDF and runs a small wasm program. 
+In order to run it for you're specific Espressif chip, edit the ['build.sh'](/product-mini/platforms/esp-idf/build.sh) file and put the correct toolchain file (see #Cross-compilation) and `IDF_TARGET`.
```
Comment (wenyongh) [inline doc/build_wamr.md:472]: In order to run it for **your** specific Espressif chip?

File: product-mini/platforms/esp-idf/CMakeLists.txt [hunk 5]
```diff
@@ -3,53 +3,57 @@
 
 # from ESP-IDF 4.0 examples/build_system/cmake/idf_as_lib
 cmake_minimum_required(VERSION 3.5)
+project(wamr_on_esp32c3)
 
-project(wamr_esp_idf C)
+enable_language(ASM)
 
-enable_language (ASM)
+set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O3")
```
Comment (wenyongh) [inline product-mini/platforms/esp-idf/CMakeLists.txt:10]: Had better use
```cmake
if (NOT CMAKE_BUILD_TYPE)
  set(CMAKE_BUILD_TYPE Release)
endif ()
```
So by default it is `-O3` and developer can use `cmake -DCMAKE_BUILD_TYPE=Debug/Minimal` to set it to others.

File: product-mini/platforms/esp-idf/CMakeLists.txt [hunk 6]
```diff
@@ -3,53 +3,57 @@
...
 set(WAMR_BUILD_PLATFORM "esp-idf")
-set(WAMR_BUILD_TARGET "XTENSA")
 set(WAMR_BUILD_INTERP 1)
-set(WAMR_BUILD_FAST_INTERP 1)
-set(WAMR_BUILD_AOT 1)
+set(WAMR_BUILD_FAST_INTERP 0)
+set(WAMR_BUILD_AOT 0)
 set(WAMR_BUILD_LIBC_BUILTIN 1)
-set(WAMR_BUILD_LIBC_WASI 0)
+set(WAMR_BUILD_APP_FRAMEWORK 0)
```
Comment (wenyongh) [inline product-mini/platforms/esp-idf/CMakeLists.txt:43]: Had better use
```cmake
if (NOT DEFINED WAMR_BUILD_XXX)
  set (WAMR_BUILD_XXX 1/0)
endif ()
```
instead, so developer can use `cmake -DWAMR_BUILD_XXX=1 or 0` to control the features.

File: product-mini/platforms/esp-idf/main.c [hunk 7]
```diff
@@ -2,38 +2,109 @@
...
  * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
  */
-
+#include <stdio.h>
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "wasm_export.h"
+#include "aot_export.h"
 #include "bh_platform.h"
+#include "platform_api_vmcore.h"
```
Comment (wenyongh) [inline product-mini/platforms/esp-idf/main.c:11]: Don't include platform_api_vmcore.h, it should have been included when bh_platform.h is included.

File: product-mini/platforms/esp-idf/main.c [hunk 8]
```diff
@@ -2,38 +2,109 @@
...
-wamr_task(void *param)
+void *
+iwasm_main(void *arg)
 {
-    os_printf("WAMR task started\n");
+    (void)arg; /* unused */
+
+    printf("about to set up the stuff\n");
+
+    // setup variables for instantiating and running the wasm module
```
Comment (wenyongh) [inline product-mini/platforms/esp-idf/main.c:32]: Had better use `/* */` to add comments.

File: product-mini/platforms/esp-idf/main.c [hunk 9]
```diff
@@ -2,38 +2,109 @@
...
+    // load WASM byte buffer from byte buffer of include file
+    printf("use an internal test file, that's going to output Hello World\n");
+    wasm_file_buf = (uint8_t *)wasm_test_file;
+    wasm_file_buf_size = sizeof(wasm_test_file);
+
+    // load WASM module
+    if (!(wasm_module = wasm_runtime_load(wasm_file_buf, wasm_file_buf_size,
+                                          error_buf, sizeof(error_buf)))) {
+        printf("Error in wasm_runtime_load: %s\n", error_buf);
+        goto fail2;
```
Comment (wenyongh) [inline product-mini/platforms/esp-idf/main.c:64]: In label fail2, it unloads the wasm_module, but here the wasm_module isn't created yet, it should invalid.

File: product-mini/platforms/esp-idf/main.c [hunk 10]
```diff
@@ -2,38 +2,109 @@
...
+    }
+
+    printf("about to call wasm_runtime_instantiate\n");
+    if (!(wasm_module_inst =
+              wasm_runtime_instantiate(wasm_module, 32 * 1024, // stack size
+                                       32 * 1024,              // heap size
+                                       error_buf, sizeof(error_buf)))) {
+        printf("Error while instantiating: %s\n", error_buf);
+        goto fail1;
```
Comment (wenyongh) [inline product-mini/platforms/esp-idf/main.c:73]: Same as above, module_inst isn't created, but fail1 destroys the module_inst.

File: product-mini/platforms/esp-idf/main.c [hunk 11]
```diff
@@ -2,38 +2,109 @@
...
+    wasm_runtime_deinstantiate(wasm_module_inst);
+
+fail2:
+    // unload the module
+    printf("wasm_runtime_unload\n");
+    wasm_runtime_unload(wasm_module);
+
+    // destroy runtime environment
+    printf("wasm_runtime_destroy\n");
+    wasm_runtime_destroy();
```
Comment (wenyongh) [inline product-mini/platforms/esp-idf/main.c:92]: Seems had better to be:
```C
    // destroy the module instance
    printf("wasm_runtime_deinstantiate\n");
    wasm_runtime_deinstantiate(wasm_module_inst);

fail2:
    // unload the module
    printf("wasm_runtime_unload\n");
    wasm_runtime_unload(wasm_module);

fail1:
    // destroy runtime environment
    printf("wasm_runtime_destroy\n");
    wasm_runtime_destroy();
```

And in line 64, `goto fail1`, in line 73, `goto fail2`

File: product-mini/platforms/esp-idf/main.c [hunk 12]
```diff
@@ -2,38 +2,109 @@
...
-app_exit(uint32_t id)
+void
+app_main(void)
 {
-    os_printf("WAMR exit, id: %d\n", id);
-}
+    pthread_t t;
+    int res;
+
+    res = pthread_create(&t, NULL, iwasm_main, (void *)NULL);
```
Comment (wenyongh) [inline product-mini/platforms/esp-idf/main.c:103]: No need to create new thread to run the wasm app, runtime can run in main thread. Creating new thread is also OK, but it increases the footprint.

**Comment by wenyongh** [comment]
@1c3t3a Thanks a lot for enabling the platform esp-idf!

File: core/shared/platform/esp-idf/espidf_malloc.c [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+#include "platform_api_extension.h"
+
+void *
+os_malloc(unsigned size)
+{
+    void *buf_origin;
+    void *buf_fixed;
+    uint32 *addr_field;
+
+    buf_origin = malloc(size + 8 + sizeof(uint32));
```
[the remark this answers is not in the document]
Comment (1c3t3a) [reply core/shared/platform/esp-idf/espidf_malloc.c:11]: > Can the return addr of malloc be 8-byte aligned in esp-idf? If yes, no need to allocate extra bytes.

No apparently not.

File: core/shared/platform/esp-idf/espidf_malloc.c [hunk 14]
```diff
@@ -0,0 +1,66 @@
+#include "platform_api_vmcore.h"
+#include "platform_api_extension.h"
+
+void *
+os_malloc(unsigned size)
+{
+    void *buf_origin;
+    void *buf_fixed;
+    uint32 *addr_field;
```
[the remark this answers is not in the document]
Comment (1c3t3a) [reply core/shared/platform/esp-idf/espidf_malloc.c:9]: You mean `uintptr_t *addr_field` don't you?

File: core/shared/platform/esp-idf/espidf_platform.c [hunk 15]
```diff
@@ -6,37 +6,20 @@
...
 {
-    int ret = 0;
-    va_list ap;
-
-    va_start(ap, format);
-    ret += vprintf(format, ap);
-    va_end(ap);
-
-    return ret;
+    return printf(format);
```
[the remark this answers is not in the document]
Comment (1c3t3a) [reply core/shared/platform/esp-idf/espidf_platform.c:22]: Ah I See, got it now!

File: core/shared/platform/esp-idf/espidf_platform.c [hunk 16]
```diff
@@ -45,76 +28,20 @@ os_vprintf(const char *format, va_list ap)
...
+    return (uint64)esp_timer_get_time();
 }
 
-void
-os_munmap(void *addr, size_t size)
+uint8 *
+os_thread_get_stack_boundary(void)
 {
-    BH_FREE(addr);
+    return NULL;
```
[the remark this answers is not in the document]
Comment (1c3t3a) [reply core/shared/platform/esp-idf/espidf_platform.c:40]: I am going to look it up, but I didn't find one before. Although I am quite sure that it should exist somewhere!

File: product-mini/platforms/esp-idf/main.c (hunk 12 above)
Comment (1c3t3a) [reply product-mini/platforms/esp-idf/main.c:103]: The problem is that ESP-IDF does not accept a `pthread_self` call if not started in a new thread. The current code uses it quite often. So I added the new thread.

File: core/shared/platform/esp-idf/espidf_thread.c (hunk 1 above)
Comment (1c3t3a) [reply core/shared/platform/esp-idf/espidf_thread.c:155]: Ah you're right... that is weird, I've changed the implementation.

File: core/shared/platform/esp-idf/espidf_platform.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/shared/platform/esp-idf/espidf_platform.c:47]: Got it, so let's merge the patch firstly, and fix it in the future.

File: core/shared/platform/esp-idf/espidf_malloc.c [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    return ptr;
+}
+
+void
+os_free(void *ptr)
+{
+    void *mem_origin;
+    uint32 *addr_field;
```
Comment (wenyongh) [inline core/shared/platform/esp-idf/espidf_malloc.c:63]: Should be `uintptr *` or `void **`

File: product-mini/platforms/esp-idf/main.c [hunk 18]
```diff
@@ -2,38 +2,108 @@
  * Copyright (C) 2019 Intel Corporation.  All rights reserved.
  * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
  */
-
+#include <stdio.h>
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "wasm_export.h"
+#include "aot_export.h"
```
Comment (xujuntwt95329) [inline product-mini/platforms/esp-idf/main.c:9]: Seems we don't need the `aot_export.h`

File: product-mini/platforms/esp-idf/main.c [hunk 19]
```diff
@@ -2,38 +2,108 @@
...
+    if (!(wasm_module_inst =
+              wasm_runtime_instantiate(wasm_module, 32 * 1024, // stack size
+                                       32 * 1024,              // heap size
+                                       error_buf, sizeof(error_buf)))) {
+        printf("Error while instantiating: %s\n", error_buf);
+        goto fail2;
+    }
+
+    printf("run main() of the application\n");
+    void *ret = app_instance_main(wasm_module_inst);
```
Comment (xujuntwt95329) [inline product-mini/platforms/esp-idf/main.c:76]: Had better define the variable at the beginning of the function.

**Comment by wenyongh** [comment]
@1c3t3a I did several minor changes according to @xujuntwt95329's comments and merged the patch. And more discussion for the commit:
- For os_thread_env_init/os_thread_env_destroy, it should be unnecessary to implement them as the implementation of os_thread APIs is based on pthread.
- For AOT mode running error, it seems that it is due to that the AOT code isn't executable: os_mmap() just calls malloc.
- For os_thread_get_stack_boundary, had better implement it in the future if we can.

**Pull request merged**